lib/Storage/S3StorageClient.cpp

```diff
@@ -0,0 +1,379 @@
+}
+
+S3StorageClient::S3StorageClient(const llvm::StringRef RawURL) {
+  // Url format is:
+  // s3://<username>:<password>@<region>+<host:port>/<bucket name>/<path>
+  revng_assert(isS3URL(RawURL));
+
+  Aws::Client::ClientConfiguration Config(false, "standard", true);
+  Config.enableEndpointDiscovery = false;
+
+  llvm::StringRef URL(RawURL);
+
+  if (URL.starts_with("s3://")) {
+    Config.scheme = Aws::Http::Scheme::HTTP;
+    SerializedString += "s3://";
+    revng_assert(URL.consume_front("s3://"));
+  } else {
+    Config.scheme = Aws::Http::Scheme::HTTPS;
+    SerializedString += "s3s://";
+    revng_assert(URL.consume_front("s3s://"));
+  }
+
+  llvm::StringRef Username = consumeSplit(URL, ':');
+  llvm::StringRef Password = consumeSplit(URL, '@');
+  SerializedString += "<user>:<pwd>@";
+
+  Aws::Auth::AWSCredentials Credentials(Username.str(), Password.str());
+
+  Config.region = consumeSplit(URL, '+').str();
+  Config.endpointOverride = consumeSplit(URL, '/').str();
+  SerializedString += Config.region + '+' + Config.endpointOverride + '/';
+
+  Client = { Credentials,
+             Config,
+             Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Always,
+             false };
+
+  if (URL.contains('/')) {
+    Bucket = consumeSplit(URL, '/').str();
+    SubPath = URL.str();
+  } else {
+    Bucket = URL.str();
+    SubPath = "";
+  }
+
+  SerializedString += Bucket + '/' + SubPath;
+}
+
+llvm::Expected<std::unique_ptr<S3StorageClient>>
+S3StorageClient::fromURL(const llvm::StringRef URL) {
+  if (not SDKInit.has_value())
+    SDKInit.emplace();
+
+  auto Instance = std::make_unique<S3StorageClient>(URL);
+  std::string IndexPath = resolvePath(Instance->SubPath, IndexName);
+
+  Aws::S3::Model::GetObjectRequest Request;
+  Request.SetBucket(Instance->Bucket);
+  Request.SetKey(IndexPath);
+
+  Aws::S3::Model::GetObjectOutcome Result = Instance->Client.GetObject(Request);
+  if (not Result.IsSuccess()) {
+    using Aws::Http::HttpResponseCode::NOT_FOUND;
+    if (Result.GetError().GetResponseCode() == NOT_FOUND)
+      return Instance;
+    else
+      return toError(Result);
+  }
+
+  std::string SerializedIndex;
+  {
+    constexpr std::size_t BufSize = 4096;
+    char Buffer[BufSize];
+    auto &Body = Result.GetResult().GetBody();
+
+    while (!Body.eof()) {
+      Body.read(Buffer, BufSize);
+      SerializedIndex.append(Buffer, Body.gcount());
+    }
+  }
+
+  llvm::yaml::Input YAMLInput(SerializedIndex);
+  YAMLInput >> Instance->FilenameMap;
+
+  return Instance;
+}
+
+std::string S3StorageClient::toString() const {
+  return SerializedString;
+}
+
+llvm::Expected<bool> S3StorageClient::exists(const llvm::StringRef Path) {
+  return FilenameMap.count(Path) != 0;
+}
+
+llvm::Error S3StorageClient::createDirectory(const llvm::StringRef Path) {
+  return llvm::Error::success();
+}
+
+llvm::Error S3StorageClient::remove(const llvm::StringRef Path) {
+  FilenameMap.erase(Path);
```

Comment:
Assert it was there, IIRC `erase` fails silently.

Reply:
added